### PR TITLE
cosmos: changes to builder pattern, +replace/delete, ...

### PR DESCRIPTION
**Pull request by nayato**
also:
- removes key printing from AuthorizationToken's Debug
- minor reduction in string copies
- replaces crypto crate with md5

### CODE REVIEW

**Comment by nayato** [comment]
Looks like it's possible to get rid of unwrap() by moving to hyper 0.12 which I'll explore in a separate PR

File: src/azure/cosmos/client.rs [hunk 1]
```diff
@@ -92,45 +64,44 @@ enum ResourceType {
 }
 
 pub struct Client {
-    hyper_client: hyper::Client<hyper_tls::HttpsConnector<hyper::client::HttpConnector>>,
-    authorization_token: AuthorizationToken,
+    hyper_client: Rc<hyper::Client<HttpsConnector<hyper::client::HttpConnector>>>,
```
Comment (MindFlavor) [inline src/azure/cosmos/client.rs]: This is excellent, it makes much more sense :+1:

File: src/azure/cosmos/document_requests.rs [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+    request: Request,
+}
+
+impl DocumentRequestExt for QueryDocumentRequest {
+    fn request(&mut self) -> &mut Request {
+        &mut self.request
+    }
+}
+
+lazy_static! {
```
Comment (MindFlavor) [inline src/azure/cosmos/document_requests.rs]: I like this. I've been wondering myself how to express this better! Thank you!

File: examples/query_document00.rs [hunk 3]
```diff
@@ -50,14 +48,11 @@ fn code() -> Result<(), Box<Error>> {
 
     let client = Client::new(&core.handle(), authorization_token)?;
 
-    let options = QueryDocumentOptions::default();
-
-    let future = client.query_document_json(
+    let future = client.query_document(
```
Comment (MindFlavor) [inline examples/query_document00.rs]: It is indeed much more elegant :+1:

File: src/azure/cosmos/partition_key.rs [hunk 4]
```diff
@@ -59,3 +51,25 @@ impl<'a> IntoIterator for PartitionKey<'a> {
         }
     }
 }
+
```
Comment (MindFlavor) [inline src/azure/cosmos/partition_key.rs]: This is weird. I've run into this String &String &str dance myself but I've been unable to find a solution either :face_with_head_bandage:

**Comment by MindFlavor** [comment]
Excellent work! The example is much more elegant your way. 

Can you please fix the travis error first? I will then fix the conflicts myself (I've resumed once more the rust-crypto crate for md5 but I think your specific crate md5 is better) and merge it.

**Comment by nayato** [comment]
sorry about Travis. Should be fixed now. I also rebased to current master now. I'd suggest to hold off releasing this until move to hyper 0.12 which I'll take on once this is merged. That'll clean up the cosmos client API from Results

**Comment by nayato** [comment]
I have a first cut of hyper 0.12 migration here: https://github.com/nayato/AzureSDKForRust/pull/1/files
once this is through, I'll get to revisit it a bit further.

**Pull request merged**